src/main/java/httpserver/HttpRouter.java

```diff
@@ -6,17 +6,26 @@
 
     private final Action methodNotAllowed;
     private final Map<String, Map<String, Action>> routes;
+    private final Action badRequest;
 
-    public HttpRouter(Map<String, Map<String, Action>> routes, Action methodNotAllowed) {
-        this.routes = routes;
+    public HttpRouter(HttpRoutes routes, Action methodNotAllowed, Action badRequest) {
+        this.routes = routes.getRoutes();
         this.methodNotAllowed = methodNotAllowed;
+        this.badRequest = badRequest;
     }
 
     public HttpResponse route(HttpRequest request) {
+        if(request == null) {
```

Comment:
If we have an `Optional` where we call `route` we can use the `Optional` API to only call route if the object in the "optional" box is an `HttpRequest` which ensures that the request is never null.

https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#map-java.util.function.Function-

Reply:
Thanks. I finally got map to work.  I think I wasn't fully understanding the functionality.